File: packages/models/lib/src/widgets/features/text_style.dart
// Flutter imports:
// ignore_for_file: public_member_api_docs

import 'package:equatable/equatable.dart';
// Flutter imports:
import 'package:flutter/material.dart';
// Project imports:
import 'package:theta_models/theta_models.dart';

class FTextStyle extends Equatable {
  /// Set of funcs to use TextStyle in Teta's widgets
  const FTextStyle({
    this.fill = const FFill(
      type: FFillType.solid,
      levels: [
        FFillElement(color: 'ffffff', stop: 0),
      ],
    ),
    this.fontSize = const FFontSize(),
    this.fontFamily = 'Poppins',
    this.fontWeight = const FFontWeight(),
    this.textDecoration = const FTextDecoration(),
    this.textAlign = const FTextAlign(),
    this.fontStyle = const FFontStyle(),
    this.textStyleModel,
    this.textDirection = const FTextDirection(),
    this.letterSpacing = const FTextTypeInput(),
    this.lineSpacing = const FTextTypeInput(),
  });

  final FFill fill;
  final FFontSize fontSize;
  final String fontFamily;
  final FFontWeight fontWeight;
  final FTextDecoration textDecoration;
  final FTextAlign textAlign;
  final FFontStyle fontStyle;
  final String? textStyleModel;
  final FTextDirection textDirection;
  final FTextTypeInput letterSpacing;
  final FTextTypeInput lineSpacing;

  @override
  List<Object?> get props => [
        fill,
        fontSize,
        fontFamily,
        fontWeight,
        textDecoration,
        textAlign,
        fontStyle,
        textStyleModel,
        textDirection,
        letterSpacing,
        lineSpacing,
      ];

  static FTextStyle fromJson(final Map<String, dynamic> doc) {
    return FTextStyle(
      fill: doc[DBKeys.fill] != null
          ? FFill.fromJson(doc[DBKeys.fill] as Map<String, dynamic>)
          : const FFill(
              type: FFillType.solid,
              levels: [FFillElement(color: '000000', stop: 0)],
            ),
      fontSize: doc[DBKeys.fontSize] != null
          ? FFontSize.fromJson(doc[DBKeys.fontSize])
          : const FFontSize(),
      fontFamily: doc[DBKeys.fontFamily] as String? ??
          doc['family'] as String? ??
          'Poppins',
      fontWeight: doc[DBKeys.fontWeight] != null
          ? FFontWeight.fromJson(doc[DBKeys.fontWeight])
          : const FFontWeight(),
      textDecoration: doc[DBKeys.textDecoration] != null
          ? FTextDecoration.fromJson(doc[DBKeys.textDecoration] as String)
          : const FTextDecoration(),
      textAlign: doc[DBKeys.textAlign] != null
          ? FTextAlign.fromJson(doc[DBKeys.textAlign] as String)
          : const FTextAlign(),
      fontStyle: doc[DBKeys.fontStyle] != null
          ? FFontStyle.fromJson(doc[DBKeys.fontStyle])
          : const FFontStyle(),
      textDirection: doc[DBKeys.textDirection] != null
          ? FTextDirection.fromJson(doc[DBKeys.textDirection] as String)
          : const FTextDirection(),
      textStyleModel: doc[DBKeys.textStyleModel] as String?,
      letterSpacing: doc[DBKeys.letterSpacing] != null
          ? FTextTypeInput.fromJson(doc[DBKeys.letterSpacing])
          : const FTextTypeInput(),
      lineSpacing: doc[DBKeys.lineSpacing] != null
          ? FTextTypeInput.fromJson(doc[DBKeys.lineSpacing])
          : const FTextTypeInput(),
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
        DBKeys.fill: fill.toJson(),
        DBKeys.fontSize: fontSize.toJson(),
        DBKeys.fontFamily: fontFamily,
        DBKeys.fontWeight: fontWeight.toJson(),
        DBKeys.textDecoration: textDecoration.toJson(),
        DBKeys.textAlign: textAlign.toJson(),
        DBKeys.fontStyle: fontStyle.toJson(),
        DBKeys.textDirection: textDirection.toJson(),
        DBKeys.textStyleModel: textStyleModel,
        DBKeys.letterSpacing: letterSpacing.toJson(),
        DBKeys.lineSpacing: lineSpacing.toJson(),
      };

  /// Returns style (TextStyle) only
  /// ```dart
  /// TextStyle(
  ///   color: Colors.black,
  ///   fontWeight: FontWeight.normal,
  ///   fontSize: 16,
  ///   fontFamily: '',
  ///   fontStyle: FontStyle.normal,
  ///   decoration: TextDecoration.none,
  /// ),

  FTextStyle copyWith({
    FFill? fill,
    FFontSize? fontSize,
    String? fontFamily,
    FFontWeight? fontWeight,
    FTextDecoration? textDecoration,
    FTextAlign? textAlign,
    FFontStyle? fontStyle,
    FTextDirection? textDirection,
    String? textStyleModel,
    FTextTypeInput? letterSpacing,
    FTextTypeInput? lineSpacing,
  }) {
    return FTextStyle(
      fill: fill ?? this.fill,
      fontSize: fontSize ?? this.fontSize,
      fontFamily: fontFamily ?? this.fontFamily,
      fontWeight: fontWeight ?? this.fontWeight,
      textDecoration: textDecoration ?? this.textDecoration,
      textAlign: textAlign ?? this.textAlign,
      fontStyle: fontStyle ?? this.fontStyle,
      textDirection: textDirection ?? this.textDirection,
      textStyleModel: textStyleModel ?? this.textStyleModel,
      letterSpacing: letterSpacing ?? this.letterSpacing,
      lineSpacing: lineSpacing ?? this.lineSpacing,
    );
  }

  FTextStyle setModelNull() {
    return FTextStyle(
      fill: fill,
      fontSize: fontSize,
      fontFamily: fontFamily,
      fontWeight: fontWeight,
      textDecoration: textDecoration,
      textAlign: textAlign,
      fontStyle: fontStyle,
      textDirection: textDirection,
      textStyleModel: null,
      letterSpacing: letterSpacing,
      lineSpacing: lineSpacing,
    );
  }
}

Widget example() {
  return const Text(
    '',
    style: TextStyle(
      color: Colors.black,
      fontWeight: FontWeight.normal,
      fontSize: 16,
      fontFamily: '',
      fontStyle: FontStyle.normal,
      decoration: TextDecoration.none,
      letterSpacing: 1,
      height: 1,
    ),
    textAlign: TextAlign.center,
  );
}

File: figma_plugin/src/features/text_style.ts
import { FFill } from "./fill";
import { FFontSize } from "./font_size";
import { FFontWeight } from "./font_weight";
import { FTextDecoration, TextDecoration } from "./text_decoration";
import { FTextAlign, TextAlign } from "./text_align";
import { FFontStyle } from "./font_style";
import { FTextDirection, TextDirection } from "./text_direction";
import { FTextTypeInput } from "./text_type_input";
import { FontStyle } from "./font_style";
import { DBKeys } from "../key_constants";
import { SizeUnit } from "../features/size";

export class FTextStyle {
  fill: FFill;
  fontSize: FFontSize;
  fontFamily: string;
  fontWeight: FFontWeight;
  textDecoration: FTextDecoration;
  textAlign: FTextAlign;
  fontStyle: FFontStyle;
  textDirection: FTextDirection;
  letterSpacing: FTextTypeInput;
  lineSpacing: FTextTypeInput;
  textStyleModel: string;

  constructor({
    fill = new FFill({}),
    fontSize = new FFontSize({ size: 16, unit: SizeUnit.pixel }),
    fontFamily = "",
    fontWeight = new FFontWeight({}),
    textDecoration = new FTextDecoration(TextDecoration.none),
    textAlign = new FTextAlign(TextAlign.left),
    fontStyle = new FFontStyle({ value: FontStyle.normal }),
    textDirection = new FTextDirection({ value: TextDirection.ltr }),
    letterSpacing = new FTextTypeInput({}),
    lineSpacing = new FTextTypeInput({}),
    textStyleModel = ""
  }: {
    fill?: FFill;
    fontSize?: FFontSize;
    fontFamily?: string;
    fontWeight?: FFontWeight;
    textDecoration?: FTextDecoration;
    textAlign?: FTextAlign;
    fontStyle?: FFontStyle;
    textDirection?: FTextDirection;
    letterSpacing?: FTextTypeInput;
    lineSpacing?: FTextTypeInput;
    textStyleModel?: string;
  }) {
    this.fill = fill;
    this.fontSize = fontSize;
    this.fontFamily = fontFamily;
    this.fontWeight = fontWeight;
    this.textDecoration = textDecoration;
    this.textAlign = textAlign;
    this.fontStyle = fontStyle;
    this.textDirection = textDirection;
    this.letterSpacing = letterSpacing;
    this.lineSpacing = lineSpacing;
    this.textStyleModel = textStyleModel;
  }

  toJson(): any {
    return {
      [DBKeys.fill]: this.fill.toJson(),
      [DBKeys.fontSize]: this.fontSize.toJson(),
      [DBKeys.fontFamily]: this.fontFamily,
      [DBKeys.fontWeight]: this.fontWeight.toJson(),
      [DBKeys.textDecoration]: this.textDecoration.toJson(),
      [DBKeys.textAlign]: this.textAlign.toJson(),
      [DBKeys.fontStyle]: this.fontStyle.toJson(),
      [DBKeys.textDirection]: this.textDirection.toJson(),
      [DBKeys.letterSpacing]: this.letterSpacing.toJson(),
      [DBKeys.lineSpacing]: this.lineSpacing.toJson(),
      [DBKeys.textStyleModel]: this.textStyleModel
    };
  }
}
